pkg/config/context.go

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2023 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package config
+
+import (
+	"context"
+	"sync"
+)
+
+type configContextKey uint8
+
+const configContextKeyValue configContextKey = 0
+
+type threadSafeConfig struct {
+	sync.RWMutex
+	Config
+}
+
+// NewContext returns a new context with an empty Config.
+func NewContext() context.Context {
+	return WithConfig(Config{})
+}
+
+// NewContextWithDefaultConfig returns a new context with a default Config.
+func NewContextWithDefaultConfig() context.Context {
+	return WithConfig(Default())
+}
+
+// JoinContext returns a new context that contains a reference to the Config
+// from the specified context.
+// This function panics if the provided context does not contain a Config.
+// This function is thread-safe.
+func JoinContext(parent context.Context, contextWithConfig context.Context) context.Context {
+	obj := contextWithConfig.Value(configContextKeyValue)
+	if obj == nil {
+		panic("config is missing from context")
+	}
+	config, ok := obj.(*threadSafeConfig)
+	if !ok {
+		panic("config in context is invalid type")
+	}
+	return context.WithValue(parent, configContextKeyValue, config)
+}
+
+// WithConfig returns a new context with the provided Config.
+func WithConfig(config Config) context.Context {
+	return WithContext(context.Background(), config)
+}
+
+// WithContext sets the provided Config in the provided context.
+// If ctx is nil, one is created with context.Background.
+func WithContext(ctx context.Context, config Config) context.Context {
+	if ctx == nil {
+		ctx = context.Background()
+	}
+	return context.WithValue(
+		ctx,
+		configContextKeyValue,
+		&threadSafeConfig{Config: config})
+}
+
+// SetContext allows the Config in the provided context to be updated.
+// This function panics if the provided context does not contain a Config.
+// This function is thread-safe.
+func SetContext(ctx context.Context, setFn func(config *Config)) context.Context {
+	obj := ctx.Value(configContextKeyValue)
+	if obj == nil {
+		panic("config is missing from context")
+	}
+	config, ok := obj.(*threadSafeConfig)
+	if !ok {
+		panic("config in context is invalid type")
+	}
+	config.Lock()
+	defer config.Unlock()
+	copyOfConfig := config.Config
```

Comment:
This copy shouldn't be needed, right?

Reply:
It prevents someone from accidentally hanging onto a reference to the `Config` object stored in the context.